southworks/BotBuilder-Samples
# [VSIX Prototype] Add BotManager VSIX project

## Proposed Changes
This PR adds a Visual Studio Extension project to simplify the process of creating the resources and deploying to Azure a **_QnAMaker_** or a **_Teams_** Bot.

**_Note: This is just a prototype for a design review, the functionality will be added in another PR._**

The extension consists of two Bot Templates ([QnAMaker](https://github.com/microsoft/BotBuilder-Samples/tree/master/samples/csharp_dotnetcore/11.qnamaker) and [Teams](https://github.com/microsoft/BotBuilder-Samples/tree/master/samples/csharp_dotnetcore/57.teams-conversation-bot)) and allows the users to create a new project based on these templates.
Once the project is ready to be deployed, the extension provides a Wizard to create the necessary resources in Azure and deploy the bot in a few simple steps. All the steps can be executed using default values but the user can opt for setting the values manually.

Here are some screenshots:

The available templates, QnAMaker and Teams Bots.
![image](https://user-images.githubusercontent.com/44245136/84530333-56002a00-acb9-11ea-8235-1a138c2f2bcd.png)

The new `Manage Resources and Deploy Manager` button in the project's context menu
![image](https://user-images.githubusercontent.com/44245136/84530399-7af49d00-acb9-11ea-957b-c915051b8a9f.png)

The Bot Manager Wizard with the steps to deploy the bot in Azure.
![image](https://user-images.githubusercontent.com/44245136/84530742-d0c94500-acb9-11ea-8982-d57fa8806f4b.png)

## Testing
First, go to `{your repository directory}/generators/vsix-vs-win/BotManagerVSIX` and open the `BotManagerVSIX.sln` solution
Then, there are two alternatives

_Alternative A_

* Build the VSIX project
* Go to `bin/debug` folder under your project's directory and run the BotManagerVSIX.vsix` file to install the extension in your Visual Studio (**NOTE**: first, ensure your Visual Studio instance is closed)
* Once the extension is installed, open Visual Studio and create a new project (QnA or Teams).
* After the project is loaded, _right-click_ on it and select _Bot Resources and Deploy Manager_ in the context menu. Then check the QnA and Teams option depending on the selected template and follow the wizard to create and deploy your bot.

_Alternative B_

* Build the VSIX project
* Run the project. It will open a Visual Studio Experimental Instance
* In the Experimental instance, create a new project (QnA or Teams).
* After the project is loaded, _right-click_ on it and select _Bot Resources and Deploy Manager_ in the context menu. Then check the QnA and Teams option depending on the selected template and follow the wizard to create and deploy your bot.
